mmocr/models/textrecog/decoders/abc_recog_decoder.py

```diff
@@ -0,0 +1,108 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import random
+
+import torch
+
+from mmocr.models.builder import DECODERS
+from mmocr.models.textrecog.layers import AttentionGRUCell
+from .base_decoder import BaseDecoder
+
+
+@DECODERS.register_module()
+class ABCRecogDecoder(BaseDecoder):
+    """Decoder for ABCNet's recognition branch.
+
+    Args:
+        num_channels (int): Number of channels of hidden vectors :math:`E`.
+        num_chars (int): Number of text characters :math:`C`.
+        max_seq_len (int): Maximum text sequence length :math:`T`.
+        teacher_forcing (float): The probability of enabling the teacher
+            forcing strategy during training.
+        dropout (float): Dropout rate.
+        init_cfg (dict or list[dict], optional): Initialization configs.
+    """
+
+    def __init__(self,
+                 num_channels=None,
+                 num_chars=None,
+                 max_seq_len=None,
+                 teacher_forcing=0.5,
+                 dropout=0.1,
+                 init_cfg=dict(type='Xavier', layer='Conv2d'),
+                 **kwargs):
+        super().__init__(init_cfg=init_cfg)
+        self.num_chars = num_chars
+        self.max_seq_len = max_seq_len
+        self.teacher_forcing = teacher_forcing
+        self.attention_cell = AttentionGRUCell(num_channels, num_chars,
+                                               dropout)
+
+    def forward_train(self, feat, out_enc, targets_dict, img_metas):
+        """
+        Args:
+            feat (Tensor): Unused.
+            out_enc (Tensor): Encoder output of shape
+                :math:`(N, T_e, E)`.
+            targets_dict (dict): A dict with the key ``padded_targets``, a
+                tensor of shape :math:`(N, T)`. Each element is the index of a
+                character.
+            img_metas (dict): Unused.
+
+        Returns:
+            Tensor: A raw logit tensor of shape :math:`(N, T, C)`.
+        """
+        return self._decode(out_enc,
+                            targets_dict['padded_targets'].to(out_enc.device))
+
+    def forward_test(self, feat, out_enc, img_metas):
+        """
+        Args:
+            feat (Tensor): Unused.
+            out_enc (Tensor): Encoder output of shape
+                :math:`(N, T_e, E)`.
+            img_metas (dict): Unused.
+
+        Returns:
+            Tensor: A raw logit tensor of shape :math:`(N, T, C)`.
+        """
+        return self._decode(out_enc)
+
+    def _decode(self, out_enc, target=None):
```

Comment:
Yes, but for this module, I think it's the best way to reuse the code. They also cannot be merged back into `forward` since splitting `forward_train` and `forward_test` might be explicitly called from outside.